ta4j-core/src/test/java/org/ta4j/core/indicators/RecentSwingHighIndicatorTest.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2023 Ta4j Organization & respective
+ * authors (see AUTHORS)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.indicators;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+import org.ta4j.core.BarSeries;
+import org.ta4j.core.Indicator;
+import org.ta4j.core.mocks.MockBarSeries;
+import org.ta4j.core.num.NaN;
+import org.ta4j.core.num.Num;
+
+import java.util.function.Function;
+import static org.junit.Assert.assertEquals;
+
+import org.ta4j.core.Bar;
+import static org.ta4j.core.TestUtils.assertNumEquals;
+import org.ta4j.core.mocks.MockBar;
+
+public class RecentSwingHighIndicatorTest extends AbstractIndicatorTest<Indicator<Num>, Num> {
```

Comment:
In that case swing high becomes rather subjective, and you can choose the first, middle, or last bar of the plateau.

Reply:
True, but I was thinking also at your definition you pasted above. Imagine a scenario where the price is stable and has a big up peak, and then go back to previous stable price. You see the graph of this and you clearly think there's a swing high there. Would you say that, that is not a swing high if the price at the peak is repeated more than once (e.g. twice)? If not, to me the question become, what usefulness has this swing high definition if such (and similar) scenarios are taken out.

Maybe in the application you are imagining this is not a problem (it is in mine (zigzag), and quiet significant)? If so, can you share your application, so I can understand more?

To generalize, there is a parameter that defines the max number of ticks allowed at the top/bottom of a swing high/low, and in your case is 1, in my case might be more than 1 (how much I don't know yet :smile: )

Reply:
The code has been updated to handle flat peaks and valleys.

Reply:
from a quick look  I think your code now identify the point in the center as swing, but so it wouldn't work with an even number of ticks in the flat peak/valley.

Reply:
The loop logic is for identifying the bars surrounding the peak/valley. There could be an arbitrary number of bars within the flat peak/valley.

Reply:
That I think I understood but my objection still stands. I tried to exemplify it with a test but got some odd behavior which stopped me doing so, this test fully pass but I believe some assertions shouldn't:

```
    @Test
    public void testCalculate_PricePlateau_ReturnsValue() {
        List<Bar> bars = new ArrayList<>();
        bars.add(new MockBar(10, 10, 10, 10, numFunction));
        bars.add(new MockBar(11, 11, 11, 11, numFunction));
        bars.add(new MockBar(12, 12, 12, 12, numFunction));
        bars.add(new MockBar(12, 12, 12, 12, numFunction));
        bars.add(new MockBar(12, 12, 12, 12, numFunction));
        bars.add(new MockBar(11, 11, 11, 11, numFunction));
        bars.add(new MockBar(10, 10, 10, 10, numFunction));
        BarSeries newSeries = new MockBarSeries(bars);

        RecentSwingHighIndicator swingHighIndicator = new RecentSwingHighIndicator(newSeries, 2);

        assertNumEquals(NaN.NaN, swingHighIndicator.getValue(0));
        assertNumEquals(NaN.NaN, swingHighIndicator.getValue(1));
        assertNumEquals(NaN.NaN, swingHighIndicator.getValue(2));
        assertNumEquals(12, swingHighIndicator.getValue(3));
        assertNumEquals(12, swingHighIndicator.getValue(4));
        assertNumEquals(12, swingHighIndicator.getValue(5));
        assertNumEquals(12, swingHighIndicator.getValue(6));
        assertNumEquals(12, swingHighIndicator.getValue(7));
        assertNumEquals(12, swingHighIndicator.getValue(15557));
        assertNumEquals(12, swingHighIndicator.getValue(519812891));
    }
    ```

In this test case and based on the logic you want to follow, shouldn't only index = 3 return a value, and eveything else `NaN`?

Reply:
> True, but I was thinking also at your definition you pasted above. Imagine a scenario where the price is stable and has a big up peak, and then go back to previous stable price. You see the graph of this and you clearly think there's a swing high there. Would you say that, that is not a swing high if the price at the peak is repeated more than once (e.g. twice)? If not, to me the question become, what usefulness has this swing high definition if such (and similar) scenarios are taken out.
> 
> Maybe in the application you are imagining this is not a problem (it is in mine (zigzag), and quiet significant)? If so, can you share your application, so I can understand more?
> 
> To generalize, there is a parameter that defines the max number of ticks allowed at the top/bottom of a swing high/low, and in your case is 1, in my case might be more than 1 (how much I don't know yet 😄 )
> 
> It's like defining when a mountain has a peak, and when it becomes upland instead 😄

The key to useful technical analysis is to focus on patterns that are unambiguously signals. If you need to squint at it to make it fit ("that could be a peak") then you're not going to be doing this for long. The goal of these indicator is to identify **unambiguous** swing highs and lows with the core trading tenent of it's OK to miss opportunities because there will always be more. In other words I'm fine with errors of omission, but not of commission. Increasing scope to handle N-bar plateaus and valleys introduces inherent ambiguity and subjectivity which is undesirable. I'm going to revert these indicators back to their previous implementation.

Reply:
thanks for you detailed explanation